README.md
# Run For Money

本插件是我的第二个作品。是看了 @骑单车の茂茂 (https://space.bilibili.com/8957291) 大大的 MC版《全员逃走中》系列后制作的。

## 此插件的功能 & 特点

* 基本分队
* 管理员可以掌控游戏流程
* ~~有简单的事件机制，在 snw.rfm.api.events 包内，可以监听。~~ **现在 API 内容已被分割。 见 [RFM API](https://github.com/SNWCreations/RunForMoneyAPI) 。**
* 可配置性强，配置文件有详细的注释，便于理解。
* 猎人分组，适用于某些失败了就会放出猎人的任务，可以单独放出也可以成组放出。
* 命令丰富，对管理员友好。
* 游戏数据查询方便。

查看此插件的命令请移步 [COMMANDS.md](https://github.com/SNWCreations/RunForMoney/blob/main/COMMANDS.md) 文件。

### **注意:**
* 此插件支持的 Minecraft 版本: **1.16.X** ，作者已测试了 1.16.5 ，理论上 1.16 的所有子版本都可以使用。
* 任务需要自行实现。
* ~~**使用此插件时，需要把被用于游戏的地图重命名为 'world' (不带单括号)。**~~ 自 v1.8.0 起不再强制要求，但不提供合适的世界名会导致终止间位置不能被定义。
* 使用此插件时，需要把服务器配置的 "allow-flight" 设为 true 。
* **不可以使用 /reload 指令，否则已有游戏数据将全部丢失。**
* 如果要在一张地图内重复进行游戏，请注意记录 B币榜 的内容！因为在启动新游戏时 B币榜 的数据将会清空！

## 获取插件！

### Maven 构建

以下内容需要你已经安装了 Maven 。

自行 clone 此仓库，在目录根目录下运行命令 'mvn clean package' 即可。
构建完成后，在 target 文件夹下的 'rfm-X.X.X.jar' 即为成品。

### Releases

在右侧的 Releases 中下载！

## 关于本插件一些机制的解释

若仍有问题，可以在 Issues 中提问。

### 队伍 & 组

队伍可以理解为阵营，在本插件中只有 猎人 和 逃走队员 这两个队伍。

而 组 是为管理员更好的管理猎人设计的，只有猎人能成为某个组的一员。

并且，每个猎人只能加入一个组。

## 一些声明

此作品 指 此仓库 下的所有由本人编写的文件及其产物。

此节内容永久有效，当此节内容有更新时，此作品自动受到 新的此节的内容 保护。

如果此节内容与 GPLv3 许可证的内容有冲突，以本节的内容为准。

**此作品 是我的个人作品，不是来自《全员逃走中》节目组官方的作品，且与《全员逃走中》节目没有关联。**

此作品 使用 GPLv3 作为许可证授权给您。详情请见 此仓库 的 LICENSE 文件。我对此作品保留所有可能的权利。

如果您使用 此作品 ，这意味着您同意并愿意遵守此节以及 GPLv3 许可证的所有规定。

## 特别感谢

[tr7zw's NBT API](https://www.spigotmc.org/resources/nbt-api.7939)

[Jetbrains](https://www.jetbrains.com) 提供的 [IntelliJ IDEA](https://www.jetbrains.com/idea)

CHANGELOG.md
# Run For Money - 更新日志

注意:

* 本 CHANGELOG 所有标注的时间是对应版本完成的时间。
* 当有功能更新时 (也就是当版本号第二个数字有变时)，推荐认真看看更新日志，有些新功能的实现需要配置文件的参与，您可能需要更新配置文件。

## v1.8.4

* 游戏逻辑更新: 修复了弃权仍然会使硬币按比例处理(这下属于是弃权=被捕...)的问题。
* 游戏逻辑更新: 现在，允许多个玩家弃权，玩家数会正确减少。
* 技术性更新: 更新 RFM API 到 v1.6.0 。

## v1.8.3 (2022/4/30)

* 游戏逻辑更新: 现在，猎人被暂停时抓捕不会生效。
* 游戏逻辑更新: 修复了玩家被捕时剩余人数播报有误的问题。
* 命令更新: 修复了 /rfmtimer 命令不对无权限玩家生效 (不是不可执行) 的问题。
* 技术性更新: 现在，道具的内部名称不能为空字符串 (即 "") ，并且道具的 ItemStack 类型不可以是空气。
* 技术性更新: 修复了 TeamHolder#addEnabledHunter(Player) 方法未检查玩家是不是猎人的问题，这避免了潜在问题。

## v1.8.2 (2022/4/19)

* 游戏逻辑更新: 移除了对猎人和逃走队员的数量进行要求的功能，但是不能在没有猎人或没有逃走队员的情况下开始游戏。
* 游戏逻辑更新: 现在，断线重连的，在游戏中的玩家(指猎人、逃走队员)不会被强制切换到冒险模式 (此特性便于调试) 。
* 游戏逻辑更新: 现在，任何生物对其他生物造成的伤害均为 0 。
* 游戏逻辑更新: 现在，恢复了 /kill 命令的效果和虚空伤害。
* 游戏逻辑更新: 现在，玩家不会在游戏结束时被清除其物品。
* 游戏逻辑更新: 修复了在游戏中途加入的新玩家没有切换到冒险模式的问题。
* 游戏逻辑更新: 修复了道具在堆叠时无效的问题。
* 游戏逻辑更新: 修复了即使抓捕事件被取消，不应被捕的玩家仍然被移出队伍的问题。
* 命令更新: 增加了 /playerremaining 命令，用法见 COMMANDS.md 文件。
* 技术性更新: 现在不再需要 Maven 私服，API 工件迁移到 JitPack 。新的用法已在 README.md 中写明。v1.5.1 以前 (不包括 v1.5.1) 的版本仍然推荐从 Maven 私服下载，但不推荐使用。
* 技术性更新: 接入了 bStats 。
* 技术性更新: 优化了猎人释放倒计时 onNewSecond 方法的处理逻辑。
* 技术性更新: 修复了 GameController#respawn 方法无效的问题。

## v1.8.1 (2022/4/12)

* 游戏逻辑更新: 现在剩余时间不会因为游戏暂停而停止显示。
* 游戏逻辑更新: 修复了 GameController#respawn 方法未考虑玩家是猎人的情况的问题。
* 游戏逻辑更新: 修复了未正常显示剩余人数的问题。
* 技术性更新: 修改了一些细节。
* 技术性更新: 更新 RFM API 到 v1.5.1 。

## v1.8.0 (2022/4/7)

此版本及之后，API 更新将不再作为新的 Major 版本，而是作为 Minor 版本。

* **!功能性更新!** 增加了玩家别称支持，对录制视频的玩家更加友好。修改玩家别称请见新配置文件 nickname.yml 。
* **!功能性更新!** 增加了多世界支持，但只有一个世界可以进行游戏，且服务器只能同时运行一次游戏。
* **!功能性更新!** 增加了多国语言支持，目前仅有中文和英语，欢迎参与翻译！
* **!功能性更新!** *移除了* 隐藏玩家名的功能，因为此功能不能正常工作，且问题无法解决。若有好的解决方案欢迎发 Issue 提出！
* 游戏逻辑更新: 修改了一些细节，修复了一些问题。
* 游戏逻辑更新: 修复了玩家名称没有被隐藏的错误。
* 命令更新: 增加了 /rfmreload 命令，用于重新从硬盘加载配置。
* 命令更新: 随着多世界支持的加入，对 /rfmsettingsquery 命令进行了微小的调整。
* 命令更新: 修复了 /hunter, /runner, /leaveteam 命令批量处理的逻辑错误。
* 技术性更新: 移除 snw.rfm.config.ItemConfiguration 类。

## v1.7.2 (2022/4/3)

* 命令更新: 修复了 "暂停" 功能不能正常工作的问题。
* 技术性更新: 修复了注册道具所用的内部名称可以带空格的问题 (带空格的内部名称会导致管理员不能通过 /rfmitem 命令获得此道具) 。
* 技术性更新: 更新 Jetbrains Java Annotations 依赖库到 23.0.0 ，并且不再打包此库，因为无意义。

## v1.7.1 (2022/3/31)

* 技术性更新: 修改了一处同样 catch Throwable 的代码。
* 技术性更新: 将 Group 改为 `HashSet<String>` 的子类，并更正了相关用法，因为插件不应该存储玩家实例。
* 技术性更新: 移除了 MainTimer#setSpeedLevel 方法 (曾经打算用于使游戏变速) 。
* 技术性更新: 修改了一些细节。

## v1.7.0 (2022/3/28)

* 技术性更新: 更新 RFM API 到 v1.5.0 。
* 技术性更新: 移除 GameProcess#out 方法。更改了相关用法。
* 技术性更新: 现在，调用 ItemEventListener 的代码只会捕获 Exception ，因为 Error 是不可控的，程序不应该尝试 catch 它。
* 技术性更新: 优化一些细节。

## v1.6.1 (2022/3/28)

* 命令更新: 修复未注册 /forceresume 命令的问题。

## v1.6.0 (2022/3/26)

* 游戏逻辑更新: 修复了一些 snw.rfm.config.Preset 类的 init 方法所含有的潜在问题。
* 命令更新: **增加 /pause 命令** ，可以中途暂停游戏。
* 命令更新: _增加 /forceresume 命令，用于调试。_
* 命令更新: 正式移除 /forcestart 命令的别称 ~~(其实早在 v1.4.0 就应该移除了)~~
* 技术性更新: 将 MainTimer 中对 tasks 的遍历用 synchronized 环绕，防止可能的 ConcurrentModificationException (当遍历中途时有别的插件调用 GameController#registerRemainingTimeEvent 方法会导致这个问题)。
* 技术性更新: 现在 GameController#pause, GameController#resume 方法可以正常按照 throws 里所写的描述抛出相应的异常。(但是这两个异常将在未来被 IllegalStateException 替换)
* 技术性更新: 更新 NBT API 到 v2.9.2 。
* 技术性更新: 修改了一些小细节。

## v1.5.0 (2022/3/13)

* 技术性更新: 更新 RFM API 到 v1.4.1 ，并实现了新 API 的新增内容。

## v1.4.1 (2022/3/12)

* 命令更新: 更改了 /rfmtimer 的运行机制，现在默认所有管理员 **不会** 看到游戏倒计时。
* 技术性更新: 将 EventProcessor 中有关调用 ItemEventListener#onPlayerUseRequiredItem 方法的代码用 try/catch 环绕，保证不因为一个 Throwable 导致其他的实现失去被调用的机会。
* 技术性更新: 优化了 ScheduledRFMTaskImpl 的调用，会在调用后自动移除其实例。
* 技术性更新: 移除了 EventProcessor#onPlayerGameModeChanged 方法。
* 技术性更新: 修复了没有在特定游戏事件发生时创建相应的事件实例并 callEvent 的错误。
* 技术性更新: 修复了未覆盖 GameStopEvent 导致在调用 getWinner 方法时引发 UnsupportedOperationException 的错误。

## v1.4.0 (2022/3/9)

* **!功能性更新!** 现在，游戏进行时，玩家看不到其他玩家的名字 (队友之间也看不到) 。相应的命令也进行了修改。
* 命令更新: 增加了 /rfmtimer 命令，可以让管理员在游戏时看到剩余时间。
* 命令更新: 修复了 /leaveteam 命令在批量处理玩家时未使玩家从所在组离开的错误。
* 命令更新: 修复了 /leaveteam 命令判断玩家是否在队伍内的表达式不正确的错误。
* 命令更新: 移除了 /forcestart 命令的别称。
* 技术性更新: 将 HunterReleaseTimer 下的内部类 SendingActionBarMessage 移至新类 ，现在其位于 snw.rfm.util 包下。

## v1.3.2 (2022/3/7)

* 技术性更新: 修复了内部代码调用 ScheduledRFMTask#executeItNow 方法后 List.remove 方法造成 ConcurrentModficationException 导致误报错误的问题。

## v1.3.1 (2022/3/7)

* 游戏逻辑更新: 现在玩家不会受到任何伤害。
* 技术性更新: 重写了 ScheduledRFMTaskImpl 及相关用法。

## v1.3.0 (2022/3/5)

因为 API 更新了，所以我更新了！

* 游戏逻辑更新: 修复了配置文件提供的 每秒增加的B币数量 的值可以为非正数的错误，现在当提供非正数时插件将使用默认值 100 。
* 命令更新: 随着 RFM API 更新到 v1.3.0 ，rfmrespawn 命令的内部实现转移到 GameController 的实现下，并更改了相关用法。
* 技术性更新: 将 CoinTimer 重命名为 MainTimer 。
* 技术性更新: **更新 RFM API 到 v1.3.4 ，并实现了新 API 的新增内容。**

## v1.2.3 (2022/3/3)

* 命令更新: 修复了 coinlist 命令显示的 B币榜 数据没有被排序的问题。

## v1.2.2 (2022/3/3)

* 游戏逻辑更新: 修复了未正确计算游戏时间导致游戏过早结束的错误。 ~~(30分钟变成30秒了，没乘以60)~~
* 游戏逻辑更新: 修复了在猎人释放前使用猎人暂停卡可以使猎人提前移动的错误。
* 游戏逻辑更新: 修复了自某次更新以后 B币计数器 未能正常运行的错误。
* 游戏逻辑更新: 修复了 B币计数器 未正常计数的错误。 **但现在通过插件主类获取的 B币榜 数据 Map 不再是被排序过的。**
* 命令更新: 修复了 ah 命令可以对已经被启用的猎人进行操作的错误。
* 命令更新: 修复了 dh 命令可以对已经被禁用的猎人进行操作的错误。
* 技术性更新: 移除了 DelayedTimer 。更改了相关用法。

## v1.2.1 (2022/3/2)

* **!道具更新!** 移除了 保命符 (曾用内部名称: sl) 。因为从未使用，并且不完善。 ~~(其实是我懒得写)~~
* 技术性更新: 修复未对 HunterCatchPlayerEvent#isCancelled 方法的返回值做出处理的错误。
* 技术性更新: 现在 GameProcess#addTimer 方法不会检查定时器类型，因为考虑到 DelayedTimer 这个对其他 BaseCountDownTimer 的子类的包装。

## v1.2.0 (2022/2/27)

API 版本: v1.2.0

修复了一些严重错误，强烈建议更新！

* **!功能性更新!** 增加了复活功能，详情见以下日志。
* 游戏逻辑更新: 在配置文件内增加了名为 "stop_game_on_no_runner_alive" 和 "coin_multiplier_on_be_catched" 的项，详情见新配置文件。
* 游戏逻辑更新: 修复了 游戏结束时因为内部代码使用 Bukkit.getScheduler()#cancelTasks 方法后又遍历调用 BukkitRunnable#cancel 方法导致抛出异常 的问题。**若未修复此问题，会导致游戏不能正常停止，从而导致不能启动新游戏。**
* 游戏逻辑更新: 修复了上个版本的猎人暂停卡不工作的问题。
* 游戏逻辑更新: 改进了猎人暂停卡的工作方式。 ~~(这个似乎更偏技术一点)~~
* 游戏逻辑更新: 修复了一些潜在问题。
* 游戏逻辑更新: 现在，游戏结束时逃走成功的玩家将不会被传送到终止间。
* 命令更新: **增加了 /rfmrespawn 命令** ，可以让一个玩家重新加入逃走中游戏(其实从未在游戏中的玩家也可以)。
* 命令更新: 增加了 /rfmsettingsquery 命令 (别称 /rsq)，可以查询现有的逃走中游戏配置。
* 命令更新: 向 /start 命令增加了名为 countDownTime 的参数，可以控制猎人释放时间。
* 技术性更新: 重命名并重写了 GameProcess 类的 letHunterCannotMove 方法，现在此方法名为 setHunterNoMoveTime 。
* 技术性更新: 修复了已经结束的游戏进程对应的 GameController 实例仍然可以用 RunForMoney#getGameController 方法得到所造成的可能产生的问题。
* 技术性更新: 修复了更新检查器判断版本号时不严谨的问题。

src/main/java/snw/rfm/RunForMoney.java
/**
 * This file is part of RunForMoney.
 *
 * RunForMoney is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * RunForMoney is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with RunForMoney. If not, see <https://www.gnu.org/licenses/>.
 */

package snw.rfm;

import de.tr7zw.changeme.nbtapi.NBTItem;
import org.bstats.bukkit.Metrics;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.command.PluginCommand;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.Damageable;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import snw.rfm.api.GameController;
import snw.rfm.commands.*;
import snw.rfm.commands.admin.*;
import snw.rfm.commands.debug.ForceResumeCommand;
import snw.rfm.commands.debug.ForceStartCommand;
import snw.rfm.commands.group.*;
import snw.rfm.commands.hunter.ActivateHunterCommand;
import snw.rfm.commands.hunter.DeactivateHunterCommand;
import snw.rfm.commands.team.HunterCommand;
import snw.rfm.commands.team.LeaveTeamCommand;
import snw.rfm.commands.team.RunnerCommand;
import snw.rfm.config.GameConfiguration;
import snw.rfm.config.Preset;
import snw.rfm.game.GameProcess;
import snw.rfm.game.TeamHolder;
import snw.rfm.processor.EventProcessor;
import snw.rfm.processor.ExitingPickaxeProcessor;
import snw.rfm.processor.HunterPauseCardProcessor;
import snw.rfm.tasks.Updater;
import snw.rfm.util.LanguageSupport;
import snw.rfm.util.NickSupport;
import snw.rfm.util.PlaceHolderString;

import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

public final class RunForMoney extends JavaPlugin {
    private static RunForMoney INSTANCE;
    private GameProcess gameProcess;
    private GameController gameController;
    private final Map<String, Double> coinEarned = new HashMap<>();

    @Override
    public void onLoad() {
        saveDefaultConfig();
        saveResource("presets.yml", false);
        saveResource("nickname.yml", false); // 2022/4/3 NickSupport!
    }

    @Override
    public void onEnable() {
        // Plugin startup logic
        INSTANCE = this; // 2022/1/29 把 INSTANCE 引用提前，便于 Util 操作实例。
        new Metrics(this, 14980); // 2022/4/19 Hello bStats!

        LanguageSupport.loadLanguage(getConfig().getString("language", "zh_CN"));

        ConsoleCommandSender consoleSender = Bukkit.getConsoleSender();
        PluginManager pluginManager = Bukkit.getPluginManager();

        consoleSender.sendMessage("[RunForMoney] " + ChatColor.GREEN + "============ Run FOR Money ============");
        consoleSender.sendMessage("[RunForMoney] " + ChatColor.GREEN + LanguageSupport.getTranslation("setup.author_info"));

        Logger ll = getLogger();
        ll.info(LanguageSupport.getTranslation("setup.load_data"));
        GameConfiguration.init(); // 2022/2/7 v1.1.5 GameConfiguration 不应该是需要实例化的。
        Preset.init();
        NickSupport.init(); // v1.8.0 NickSupport!
        EventProcessor.init();
        TeamHolder.getInstance().init();

        registerInternalItems();

        ll.info(LanguageSupport.getTranslation("setup.register_command"));
        // region 注册命令
        registerCommand("start", new StartCommand());
        registerCommand("forcestop", new ForceStopCommand());
        registerCommand("hunter", new HunterCommand());
        registerCommand("runner", new RunnerCommand());
        registerCommand("leaveteam", new LeaveTeamCommand());
        registerCommand("endroom", new EndRoomCommand());
        registerCommand("newgroup", new NewGroupCommand());
        registerCommand("removegroup", new RemoveGroupCommand());
        registerCommand("activategroup", new ActivateGroupCommand());
        registerCommand("deactivategroup", new DeactivateGroupCommand());
        registerCommand("joingroup", new JoinGroupCommand());
        registerCommand("leavegroup", new LeaveGroupCommand());
        registerCommand("activatehunter", new ActivateHunterCommand());
        registerCommand("deactivatehunter", new DeactivateHunterCommand());
        registerCommand("resume", new ResumeCommand());
        registerCommand("grouplist", new GroupListCommand());
        registerCommand("teamlist", new TeamListCommand());
        registerCommand("coinlist", new CoinListCommand());
        registerCommand("rfmitem", new RFMItemCommand());
        registerCommand("exportcoinlist", new ExportListCommand());
        registerCommand("rfmrespawn", new RFMRespawnCommand());
        registerCommand("rfmsettingsquery", new RFMSettingsQueryCommand());
        registerCommand("rfmtimer", new RFMTimerCommand());
        registerCommand("pause", new PauseCommand());
        registerCommand("rfmreload", new RFMReloadCommand());
        registerCommand("playerremaining", new PlayerRemainingCommand());

        RFMTeamCommand.register();
        RFMGameCommand.register();
        RFMDataCommand.register();
        RFMGroupCommand.register();

        // endregion

        // region 注册调试命令
        // 警告: 以下注册的命令不应该被最终用户使用。
        registerCommand("forcestart", new ForceStartCommand());
        registerCommand("forceresume", new ForceResumeCommand());
        // endregion

        ll.info(LanguageSupport.getTranslation("setup.register_event_processor"));
        pluginManager.registerEvents(new EventProcessor(), this);
        pluginManager.registerEvents(new ExitingPickaxeProcessor(), this);

        getLogger().info(LanguageSupport.getTranslation("setup.complete"));

        if (getConfig().getBoolean("check_update", false)) { // 检查更新
            new Updater().start();
        }
    }

    @Override
    public void onDisable() {
        // Plugin shutdown logic
        if (getGameProcess() != null) {
            getLogger().info(LanguageSupport.getTranslation("unload.forcestop"));
            getGameProcess().stop();
        }
        Bukkit.getScheduler().cancelTasks(this); // make sure no tasks still running
    }

    public static RunForMoney getInstance() {
        return INSTANCE;
    }

    public GameProcess getGameProcess() {
        return gameProcess;
    }

    public void setGameProcess(@Nullable GameProcess process) {
        gameProcess = process;
    }

    public GameController getGameController() {
        return gameController;
    }

    public void setGameController(@Nullable GameController gameController) {
        this.gameController = gameController;
    }

    public Map<String, Double> getCoinEarned() {
        return coinEarned;
    }

    public void registerInternalItems() {
        // region 弃权镐
        ItemStack ep = new ItemStack(Material.WOODEN_PICKAXE);
        ItemMeta epmeta = ep.getItemMeta();
        //noinspection ConstantConditions
        epmeta.setDisplayName(ChatColor.RED + "" + ChatColor.BOLD + LanguageSupport.getTranslation("item.ep"));
        Damageable converted_meta = (Damageable) epmeta;
        converted_meta.setDamage(58);
        ep.setItemMeta((ItemMeta) converted_meta);
        NBTItem item = new NBTItem(ep);
        item.getStringList("CanDestroy").add(RunForMoney.getInstance().getConfig().getString("exiting_pickaxe_minable_block", "minecraft:diamond_block"));
        ItemStack EXITING_PICKAXE = item.getItem();
        ItemRegistry.registerItem("ep", EXITING_PICKAXE);
        // endregion

        // region 猎人暂停卡 (2022/1/30)
        Material hpctype = Material.matchMaterial(RunForMoney.getInstance().getConfig().getString("hpc_type", "minecraft:gold_ingot"));
        if (hpctype == null) {
            Bukkit.getConsoleSender().sendMessage("[RunForMoney] " + ChatColor.YELLOW +
                    new PlaceHolderString(
                            LanguageSupport.getTranslation("setup.config.item.invalid_type")).replaceArgument("itemName", LanguageSupport.getTranslation("item.hpc")));
        } else {
            ItemStack hpc = new ItemStack(hpctype);
            ItemMeta hpcmeta = hpc.getItemMeta();
            assert hpcmeta != null;
            hpcmeta.setDisplayName(ChatColor.GOLD + "" + ChatColor.BOLD + LanguageSupport.getTranslation("item.hpc"));
            hpc.setItemMeta(hpcmeta);
            HunterPauseCardProcessor hpcp = new HunterPauseCardProcessor();
            ItemRegistry.registerItem("hpc", hpc, hpcp);
            Bukkit.getPluginManager().registerEvents(hpcp, this);
        }
        // endregion
    }

    private void registerCommand(@NotNull String cmdName, @NotNull CommandExecutor executor) {
        PluginCommand cmd = getCommand(cmdName);
        if (cmd == null) {
            getLogger().severe(
                    new PlaceHolderString(LanguageSupport.getTranslation("setup.command_not_found")).replaceArgument("cmdName", cmdName).toString()
            );
            throw new NullPointerException();
        } else {
            cmd.setExecutor(executor);
        }
    }

}
